### PR TITLE
Add metadata to Session objects

### PR DESCRIPTION
**Pull request by alixdamman**
Needs changelog

### CODE REVIEW

File: larray/core/session.py [hunk 1]
```diff
@@ -873,8 +946,9 @@ def elem_equal(e1, e2):
 
     array_equals = renamed_to(element_equals, 'array_equals')
 
+    # TODO: include metadata?
```
Comment (gdementen) [inline larray/core/session.py]: This could be an option. But we need a default value and I am unsure about it. Do you have an idea?

File: larray/core/session.py [hunk 2]
```diff
@@ -1091,7 +1165,7 @@ def summary(self, template=None):
                 - for groups: 'key', 'name', 'axis_name', 'labels' and 'length',
                 - for axes: 'key', 'name', 'labels' and 'length',
                 - for arrays: 'key', 'axes_names', 'shape', 'dtype' and 'title',
-                - for all other types: 'key', 'value'.
+                - for metadata: 'key', 'value'.
```
Comment (gdementen) [inline larray/core/session.py]: what about other kinds of objects?

File: larray/core/session.py [hunk 3]
```diff
@@ -1150,9 +1231,15 @@ def summary(self, template=None):
             template[Group] = "{key}: {axis_name}{labels} >> {name} ({length})"
         if LArray not in template:
             template[LArray] = "{key}: {axes_names} ({shape}) [{dtype}]\n    {title}"
+        if Metadata not in template:
+            template[Metadata] = "{key} (metadata): {value}"
 
         def display(k, v):
-            t = Group if isinstance(v, Group) else type(v)
+            if not isinstance(v, (Axis, Group, LArray)):
```
Comment (gdementen) [inline larray/core/session.py]: I would **much** rather have a single template for Metadata objects, so that users can chose to indent it more than the rest or not to display it at all. It would also (re-) allow user-defined objects in the Session.

File: larray/inout/excel.py [hunk 4]
```diff
@@ -249,13 +256,26 @@ def _dump_item(self, key, value, *args, **kwargs):
...
+    def _dump_metadata(self, metadata):
+        if len(metadata) > 0:
+            s = _metadata_to_series(metadata)
+            s.to_excel(self.handle, '__metadata__', engine='xlsxwriter')
+
     def save(self):
         if len(self.axes) > 0:
             df = _axes_to_df(self.axes.values())
-            df.to_excel(self.handle, '__axes__', engine='xlsxwriter')
+            df.to_excel(self.handle, '__axes__', index=False, engine='xlsxwriter')
```
Comment (gdementen) [inline larray/inout/excel.py]: is this change really related to session metadata???

File: larray/inout/pandas.py [hunk 5]
```diff
@@ -291,9 +293,27 @@ def _series_to_group(series, axis):
     return LGroup(key=series.values, name=name, axis=axis)
 
 
-# ################################## #
-#    DATAFRAME <--> AXES, GROUPS     #
-# ################################## #
+def _metadata_to_series(metadata):
+    s = pd.Series(data=list(metadata.values()), index=metadata.keys(), name='')
```
Comment (gdementen) [inline larray/inout/pandas.py]: Why don't you use a 1D object LArray for this?

File: larray/inout/pandas.py [hunk 6]
```diff
@@ -291,9 +293,27 @@ def _series_to_group(series, axis):
...
+    s.index.name = 'METADATA'
+    return s
+
+
+def _series_to_metadata(series):
+
+    def _convert_value(key, value):
```
Comment (gdementen) [inline larray/inout/pandas.py]: this function does not seem to use the key argument. Remove it.

File: larray/core/session.py [hunk 7]
```diff
@@ -172,13 +202,21 @@ def __delitem__(self, key):
         del self._objects[key]
 
     def __getattr__(self, key):
-        if key in self._objects:
+        if key is 'meta':
```
[the remark this answers is not in the document]
Comment (alixdamman) [reply larray/core/session.py]: I started by trying to implement ``meta`` as a property but I faced some difficulties to make it work with Python 2.7

File: larray/tests/test_session.py [hunk 8]
```diff
@@ -469,11 +495,17 @@ def test_rdiv(session):
...
-def test_pickle_roundtrip(session):
+def test_pickle_roundtrip(session, meta):
     original = session.filter(kind=LArray)
+    original.meta = meta
     s = pickle.dumps(original)
     res = pickle.loads(s)
     assert res.equals(original)
+    if PY2:
+        from larray.core.metadata import Metadata
+        assert Metadata(res.meta) == meta
```
[the remark this answers is not in the document]
Comment (alixdamman) [reply larray/tests/test_session.py]: I'm not proud of it. After trying to implement ``meta`` in Session as a property and facing difficulties with Python 2.7 I just gave up for this part. I will change that.

File: larray/inout/hdf.py [hunk 9]
```diff
@@ -165,5 +166,12 @@ def _dump_item(self, key, value, *args, **kwargs):
         else:
             raise TypeError()
 
+    def _read_metadata(self):
+        attrs = self.handle.get_node('')._v_attrs
```
[the remark this answers is not in the document]
Comment (alixdamman) [reply larray/inout/hdf.py]: I didn't found any other way to add 'global' metadata to the HDF5 file (see https://www.pytables.org/usersguide/tutorials.html#setting-and-getting-user-attributes for more details)

File: larray/core/session.py [hunk 10]
```diff
@@ -262,9 +315,10 @@ def load(self, fname, names=None, engine='auto', display=False, **kwargs):
             handler = handler_cls(fname, kwargs['sep'])
         else:
             handler = handler_cls(fname)
-        objects = handler.read_items(names, display=display, **kwargs)
+        metadata, objects = handler.read_items(names, display=display, **kwargs)
```
[the remark this answers is not in the document]
Comment (alixdamman) [reply larray/core/session.py]: I load metadata from `read_items()` to not call `self._open_for_read()` and `self.close()` twice.

File: larray/inout/excel.py (hunk 4 above)
Comment (alixdamman) [reply larray/inout/excel.py]: Nope, sorry.

File: larray/inout/excel.py [hunk 11]
```diff
@@ -190,14 +193,14 @@ def _load_axes_and_groups(self):
         # load all axes
         sheet_axes = '__axes__'
         if sheet_axes in self.handle.sheet_names:
-            df = pd.read_excel(self.handle, sheet_axes)
+            df = pd.read_excel(self.handle, sheet_axes, index_col=None)
```
[the remark this answers is not in the document]
Comment (alixdamman) [reply larray/inout/excel.py]: Nope, sorry.

File: larray/core/session.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (gdementen) [reply larray/core/session.py]: Yes, I understand why you did this and it was the best move without modifying too much the FileHandler API. It's just that I think we can do better if we change it. Now, this is no big deal while that API is purely internal but I think it could be useful to make it public some day and in that case, the current API is not good enough.

[the remark this answers is not in the document]
Comment (alixdamman) [reply larray/core/session.py]: I agree

File: larray/core/session.py (hunk 1 above)
Comment (alixdamman) [reply larray/core/session.py]: IMO metadata should not be tested in equals. There are not properly items of a Session.

Comment (gdementen) [reply larray/core/session.py]: ok, let's go for this. Should be documented explicitly though.

File: larray/inout/pandas.py (hunk 5 above)
Comment (alixdamman) [reply larray/inout/pandas.py]: Because LArray objects are converted to pandas objects before to be dumped or loaded. I think using 1D array here instead of a pandas Series will add an unnecessary layer. Especially since we need to iterate over each element at reading to convert each metadata to the right type.

Comment (gdementen) [reply larray/inout/pandas.py]: This might be a case of premature optimization. Passing via LArray has negligible cost in my working copy (disclaimer: I modified .series slightly to go faster for 1D and I had already a few speed optimizations all over the place in my working copy, so this might not be the case on your computer).

```python
>>> metadata = {'title': 'yad aez dgrz jgzeigfejzi', 'date': datetime.now()}
>>> %timeit l = LArray(list(metadata.values()), Axis(metadata.keys())); s = l.series; s.name = ''
1.22 ms ± 57.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
>>> %timeit s = pd.Series(data=list(metadata.values()), index=metadata.keys(), name='')
1.3 ms ± 65.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Besides, I think that it might be a case where avoiding pandas structures altogether would be faster. Going via pandas is a net win for reading .csv files because their parser is so fast, but for anything else, we only gain convenience, not performance. This is probably especially true when we go via xlwings, which in the end must transform everything to "simple" Python types. Anyway, you can leave this as-is, but please at least add a comment that this is done for performance reasons.

File: larray/inout/pandas.py [hunk 12]
```diff
@@ -291,9 +293,27 @@ def _series_to_group(series, axis):
...
+        value = str(value)
```
Comment (gdementen) [inline larray/inout/pandas.py]: Why is this line needed? It seems weird to convert numeric types to string to convert them back on the next line to numeric?

File: larray/inout/pandas.py [hunk 13]
```diff
@@ -291,9 +293,27 @@ def _series_to_group(series, axis):
...
+        value = pd.to_numeric([value], errors='ignore')[0]
```
Comment (gdementen) [inline larray/inout/pandas.py]: why do you use [value] instead of just value? pd.to_numeric seems to handle scalars just fine (even if that is undocumented)

File: larray/core/session.py (hunk 1 above)
Comment (alixdamman) [reply larray/core/session.py]: The docstring starts with "*Test if groups, axes and arrays of the current session are equal to those of another session*". It's quite explicit no?

Comment (gdementen) [reply larray/core/session.py]: Not enough I think. Users have to think for themselves that "hey metadata is not in that list". I think there should be a note somewhere stating something vaguely like: "other metadata (from .meta) is **NOT** compared" or "is ignored".

File: larray/inout/pandas.py (hunk 13 above)
Comment (alixdamman) [reply larray/inout/pandas.py]: What version of pandas do you have? If I try to pass value instead of [value], I get the following error message: TypeError: 'int' object is not subscriptable

Comment (gdementen) [reply larray/inout/pandas.py]:
```python
>>> pd.to_numeric('10', errors='ignore')
10
>>> pd.to_numeric(10, errors='ignore')
10
>>> pd.__version__
'0.20.3'
```

File: larray/core/session.py [hunk 14]
```diff
@@ -1170,14 +1273,18 @@ def display(k, v):
             else:
                 return tmpl(k, v)
 
-        return '\n'.join(display(k, v) for k, v in self.items())
+        res = ''
+        if len(self.meta) > 0:
+            res = '\n'.join(display_metadata(k, v) for k, v in self.meta.items()) + '\n'
+        res += '\n'.join(display_items(k, v) for k, v in self.items())
```
Comment (gdementen) [inline larray/core/session.py]: then here, something like:
```python
if self.meta:
    res += display_items('Metadata', v) + '\n'
res += '\n'.join(display_items(k, v) for k, v in self.items())
```

The goal being that users can produce something like the following if they want to (maybe it should be the default):
```python
Metadata: 
    title: my title
    author: John Smith
axis1: a ['a0' 'a1' 'a2'] (3)
group1 -> a01: a['a0', 'a1'] (2)
```
**Now, I am unsure it's worth it** to change what you have but having no way to distinguish a metadata item from a "normal" item clearly bothers me.

**Comment by alixdamman** [comment]
There is a bug with conda for Python 2.7 : https://github.com/conda/conda/issues/7503

File: doc/source/changes/version_0_29.rst.inc [hunk 15]
```diff
@@ -48,7 +48,7 @@ New features
 
   Warnings:
 
-  - Currently, only the HDF (.h5) file format supports saving and loading metadata.
+  - Currently, only the HDF (.h5) file format supports saving and loading array metadata.
   - Metadata are not kept when actions or methods are applied on a array
```
Comment (gdementen) [inline doc/source/changes/version_0_29.rst.inc]: s/Metadata are/Metadata is/
s/a/an/

File: doc/source/changes/version_0_29.rst.inc [hunk 16]
```diff
@@ -67,6 +67,31 @@ New features
     >>> arr.meta.title = 'array for testing'
 
 
+* allowed sessions to have metadata.
+  Like for arrays, session metadata can be accessed using the syntax ``session.meta.name``.
+  See bellow for use cases:
```
Comment (gdementen) [inline doc/source/changes/version_0_29.rst.inc]: s/bellow/below/

File: doc/source/changes/version_0_29.rst.inc [hunk 17]
```diff
@@ -67,6 +67,31 @@ New features
...
+    >>> # Python 3.6+
+    >>> s = Session(arr1=ndtest(2), arr2=ndtest(3), meta=Metadata(title='my title', author='John Smith'))
+    >>> # show/access metadata
+    >>> s.meta
+    title: my title
+    author: John Smith
+
+  Warnings:
+
+  - Excel, CSV and HDF (.h5) file formats support saving and loading session metadata.
```
Comment (gdementen) [inline doc/source/changes/version_0_29.rst.inc]: * These are all the common formats, so it is a bit weird to have this in a Warning section, but maybe putting that in a Notes section by itself would be impractical.
* I would emphasize more the difference with arrays by saying something like "Contrary to array metadata, saving and loading session metadata is supported for all current session file formats: Excel, CSV and HDF (.h5).

File: doc/source/changes/version_0_29.rst.inc [hunk 18]
```diff
@@ -67,6 +67,31 @@ New features
...
+  - Metadata are not kept when actions or methods are applied on a session
```
Comment (gdementen) [inline doc/source/changes/version_0_29.rst.inc]: s/are/is/ 😄

File: larray/core/session.py [hunk 19]
```diff
@@ -30,7 +31,18 @@ class Session(object):
...
-        Objects to add written as name=object, ...
+
+        * Objects to add written as name=object
+        * meta : list of pairs or dict or OrderedDict or Metadata
+            Metadata (title, description, author, creation_date, ...) associated with the array.
+            Keys must be strings. Values must be of type string, int, float, date, time or datetime.
+
+    Warnings
+    --------
+    Metadata are not kept when actions or methods are applied on a session
```
Comment (gdementen) [inline larray/core/session.py]: are...is...

File: larray/core/session.py [hunk 20]
```diff
@@ -1164,20 +1258,25 @@ def display(k, v):
...
+                                       shape=' x '.join(str(i) for i in v.shape), dtype=v.dtype)
                 else:
                     return tmpl.format(key=k, value=str(v))
             else:
                 return tmpl(k, v)
 
-        return '\n'.join(display(k, v) for k, v in self.items())
+        res = ''
+        if len(self.meta) > 0:
+            res = 'Metadata:\n'
```
Comment (gdementen) [inline larray/core/session.py]: Not what I had in mind, but oh well... I am tired of this...
